extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/ui/KafkaTopicClient.java

```diff
@@ -0,0 +1,273 @@
+            } else {
+                consumer.seekToBeginning(List.of(topicPartition));
+            }
+            return consumer.position(topicPartition);
+        }
+    }
+
+    public Map<Integer, Long> getPagePartitionOffset(String topicName, Collection<Integer> requestedPartitions, Order order)
+            throws ExecutionException, InterruptedException {
+        assertRequestedPartitionsExist(topicName, requestedPartitions);
+
+        var result = new HashMap<Integer, Long>();
+        for (var requestedPartition : requestedPartitions) {
+            var maxPosition = getPosition(topicName, requestedPartition, order);
+            result.put(requestedPartition, maxPosition);
+        }
+
+        return result;
+    }
+
+    private List<ConsumerRecord<Bytes, Bytes>> getConsumerRecords(String topicName, Order order, int pageSize,
+            Collection<Integer> requestedPartitions, Map<Integer, Long> start, int totalMessages) {
+        List<ConsumerRecord<Bytes, Bytes>> allPartitionsResult = new ArrayList<>();
+
+        // Requesting a full page from each partition and then filtering out redundant data. Thus, we'll ensure, we read data in historical order.
+        for (var requestedPartition : requestedPartitions) {
+            List<ConsumerRecord<Bytes, Bytes>> partitionResult = new ArrayList<>();
+            var offset = start.get(requestedPartition);
+            try (var consumer = createConsumer(topicName, requestedPartition, this.config)) {
+                // Move pointer to currently read position. It might be different per partition, so requesting with offset per partition.
+                var partition = new TopicPartition(topicName, requestedPartition);
+
+                var seekedOffset = Order.OLD_FIRST == order ? offset : Long.max(offset - pageSize, 0);
+                consumer.seek(partition, seekedOffset);
+
+                var numberOfMessagesReadSoFar = 0;
+                var keepOnReading = true;
+
+                while (keepOnReading) {
+                    var records = pollWhenReady(consumer);
+                    if (records.isEmpty())
+                        keepOnReading = false;
+
+                    for (var record : records) {
+                        numberOfMessagesReadSoFar++;
+                        partitionResult.add(record);
+
+                        if (numberOfMessagesReadSoFar >= totalMessages) {
+                            keepOnReading = false;
+                            break;
+                        }
+                    }
+                }
+                // We need to cut off result, if it was reset to 0, as we don't want see entries from old pages.
+                if (Order.NEW_FIRST == order && seekedOffset == 0 && partitionResult.size() > offset.intValue()) {
+                    partitionResult.sort(Comparator.comparing(ConsumerRecord::timestamp));
+                    partitionResult = partitionResult.subList(0, offset.intValue());
```

Comment:
it's not guaranteed to be sorted

Reply:
What I meant is when using the `getConsumerRecords` method the resulting list is sorted again.

Reply:
because here we sort locally, within particular partition, and there - under multiple partitions, the merged result
here we have to sort to cut off what is redundant for partition, and there - for page

Reply:
Ok, I didn't see that. I now realize that we are creating a consumer per topic-partition. 
Kafka consumer's unfairness between partitions really complicates things for this use case.
Do you think we can create a consumer per topic, possibly reusing it or only its client id?

Reply:
no, we can't 
imagine if a user has two tabs open, or two users access this UI for a remote app. 
we enter same sode twice, shifting the offset to different position and getting wrong response in one of the pages.

Reply:
I understand that we can't reuse it between clients, I just thought creating a consumer per topic-partition was too much. Maybe there isn't a straightforward way of consuming multiple partitions in a fair way using a single consumer. We can revisit this later if needed.